src/module.c

```diff
@@ -5108,6 +5130,149 @@ int RM_GetTimerInfo(RedisModuleCtx *ctx, RedisModuleTimerID id, uint64_t *remain
+    /* Fire the client changed handler now in case we are unloading the module
+     * and need to cleanup.  */
+    moduleNotifyUserChanged(c);
+
+    c->user = DefaultUser;
+    c->authenticated = 0;
+    freeClientAsync(c);
+}
+
+/* Cleanup all clients with an auth_ctx to prevent leaking. */
+void moduleFreeAuthenticatedClients(RedisModule *module) {
+    listIter li;
+    listNode *ln;
+    listRewind(server.clients,&li);
+    while ((ln = listNext(&li)) != NULL) {
+        client *c = listNodeValue(ln);
+        if (!c->auth_ctx) continue;
+
+        RedisModuleAuthCtx *auth_ctx = (RedisModuleAuthCtx *) c->auth_ctx;
+        if (auth_ctx->module == module) { 
+            revokeClientAuthentication(c);
+        }
+    }
+}
+
+static int authenticateClientWithUser(RedisModuleCtx *ctx, user *user, RedisModuleAuthCtx *auth_ctx) {
+    if (auth_ctx && auth_ctx->authenticated_client) {
+        /* Prevent basic misuse of the auth context */
+        return REDISMODULE_ERR;
+    }
+    
+    if (user->flags & USER_FLAG_DISABLED) {
+        return REDISMODULE_ERR;
+    }
+
+    moduleNotifyUserChanged(ctx->client);
+
+    ctx->client->user = user;
+    ctx->client->authenticated = 1;
+
+    if (auth_ctx) {
+        ctx->client->auth_ctx = auth_ctx;
+        auth_ctx->authenticated_client = ctx->client;
+        auth_ctx->module = ctx->module;
+    }
+
+    return REDISMODULE_OK;
+}
+
+/* Creates a new user that is unlinked from the main ACL user dictionary. These
+ * users behave the same way as those in ACL.c except for a few minor 
+ * differences. These users do not exist within a namespace, and handling 
+ * duplicate users is the responsibility of the calling module. These users are
+ * also not attached to the redis users dictionary, so they are not returned 
+ * via ACL LIST or GETUSER. This also means that users created here must be 
+ * updated with the SetUserACL function instead of through ACL SETUSER. */
+RedisModuleUser *RM_CreateModuleUser(const char *name) {
+    RedisModuleUser *new_user = zmalloc(sizeof(RedisModuleUser));
+    new_user->user = ACLCreateUnlinkedUser();
+
+    /* Free the temporarily assigned name to assign the new one */
+    sdsfree(new_user->user->name);
+    new_user->user->name = sdsnew(name);
+    return new_user;
+}
+
+/* Frees a given user and disconnects all of the clients that have been
+ * authenticated with it. */
+int RM_FreeModuleUser(RedisModuleUser *user) {
+    ACLFreeUserAndKillClients(user->user);
+    zfree(user);
+    return REDISMODULE_OK;
+}
+
+/* Sets the user permission of a user created through the redis module 
+ * interface. The syntax is the sam as ACL SETUSER, so refer to the 
+ * documentation in acl.c for more information. */
+int RM_SetModuleUserACL(RedisModuleUser *user, const char* acl) {
+    return ACLSetUser(user->user, acl, -1);
+}
+
+/* Authenticate the current contexts user with the provided module user. 
+ * Throws an error if the user is disabled or the AuthCtx is misued */
```

Comment:
Sure, I can make that clearer. 
I was thinking you include an AuthCtx when you want to follow up on this client either because you allocated something that needs to be freed or you want to close the client after some amount of time. 

If you don't provide in any context, NULL, you don't care and are happy letting the client continue being authenticated until the end of time.